### ISSUE TITLE
Use a new backbone model

### ISSUE THREAD

**Issue by phuongdnm** [xiuqhou/Salience-DETR#59]
### Question

@xiuqhou 
Thank you so much for sharing the wonderful research!

I want to replace the current backbone model with a new one to see if it improves performance. 
I see that the codebase already supports ResNet, Swin-L, and Focal-L. 
Can you guide me on how to add or integrate a new backbone model?

Thank you in advance!

**Comment by xiuqhou** [comment]
Hi @phuongdnm , adding a backbone to Salience-DETR is very simple! Just define a `nn.Module` following the requirements below:
1. It accepts a tensor of shape `(batch_size, num_channels, height, width)` as input
2. It returns a dict of feature maps. Each value is a tensor of shape `(batch_size, num_channels, height, width)`. The keys can be anything.
3. The Module should have a `num_channels` attribute. It is a list of `num_channels` of the returned feature maps.

```python
backbone = YourBackbone()

image = torch.randn(1, 3, 256, 256)
features = backbone(image)

print([feat.shape for feat in features.values()])
# [torch.Size([1, 48, 32, 32]), torch.Size([1, 96, 16, 16]), torch.Size([1, 192, 8, 8])]

print(backbone.num_channels)
# [48, 96, 192]
```

Then you can use it in model config:

```python
# define your own backbone
backbone = YourBackbone()

neck = ChannelMapper( ...
transformer = ...
model = SalienceDETR(
    backbone=backbone,
    neck=neck,
    transformer=transformer,
    ...
)
```

**Comment by xiuqhou** [comment]
Here is an example for backbones from `timm`. The backbone returns a list. Now we need to wrap it to return a dict and add `num_channels` attribute. You can freeze some layers manually if needed.

```python
mobilenet = timm.create_model(
    "hf_hub:timm/mobilenetv4_conv_aa_large.e230_r448_in12k_ft_in1k",
    pretrained=True,
    features_only=True,
    out_indices=(1, 2, 3),
)

class TIMMBackboneWrapper(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.num_channels = model.feature_info.channels()

    def forward(self, image: torch.Tensor):
        features = self.model(image)

        # convert from list to dict, keys can be anything, we simply use the index
        # it will be {0: features[0], 1: features[1], 2: features[3], ...}
        features = dict(enumerate(features))

        return features


backbone = TIMMBackboneWrapper(model=mobilenet)
```

Here is an example for model config:

```python

import timm
import torch
from torch import nn

from models.bricks.position_encoding import PositionEmbeddingSine
from models.bricks.post_process import PostProcess
from models.bricks.salience_transformer import (
    SalienceTransformer,
    SalienceTransformerDecoder,
    SalienceTransformerDecoderLayer,
    SalienceTransformerEncoder,
    SalienceTransformerEncoderLayer,
)
from models.bricks.set_criterion import HybridSetCriterion
from models.detectors.salience_detr import SalienceCriterion, SalienceDETR
from models.matcher.hungarian_matcher import HungarianMatcher
from models.necks.channel_mapper import ChannelMapper
from models.necks.repnet import RepVGGPluXNetwork

# mostly changed parameters
embed_dim = 256
num_classes = 91
num_queries = 900
num_feature_levels = 4
transformer_enc_layers = 6
transformer_dec_layers = 6
num_heads = 8
dim_feedforward = 2048

# instantiate model components
position_embedding = PositionEmbeddingSine(embed_dim // 2, temperature=10000, normalize=True, offset=-0.5)


####### START OF THE NEW BACKBONE ################
mobilenet = timm.create_model(
    "hf_hub:timm/mobilenetv4_conv_aa_large.e230_r448_in12k_ft_in1k",
    pretrained=True,
    features_only=True,
    out_indices=(2, 3, 4),
)

# freeze the backbone
for module in [mobilenet.blocks[0], mobilenet.blocks[1]]:
    module.eval()
    for param in module.parameters():
        param.requires_grad = False

class TIMMBackboneWrapper(nn.Module):
    def __init__(self, model):
        super().__init__()
        self.model = model
        self.num_channels = model.feature_info.channels()

    def forward(self, image: torch.Tensor):
        features = self.model(image)

        # convert from list to dict, keys can be anything, we simply use the index
        features = dict(enumerate(features))

        return features


backbone = TIMMBackboneWrapper(model=mobilenet)

###### END OF NEW BACKBONE ########

neck = ChannelMapper(
    in_channels=backbone.num_channels,
    out_channels=embed_dim,
    num_outs=num_feature_levels,
)

transformer = SalienceTransformer(
    encoder=SalienceTransformerEncoder(
        encoder_layer=SalienceTransformerEncoderLayer(
            embed_dim=embed_dim,
            n_heads=num_heads,
            dropout=0.0,
            activation=nn.ReLU(inplace=True),
            n_levels=num_feature_levels,
            n_points=4,
            d_ffn=dim_feedforward,
        ),
        num_layers=transformer_enc_layers,
    ),
    neck=RepVGGPluXNetwork(
        in_channels_list=neck.num_channels,
        out_channels_list=neck.num_channels,
        norm_layer=nn.BatchNorm2d,
        activation=nn.SiLU,
        groups=4,
    ),
    decoder=SalienceTransformerDecoder(
        decoder_layer=SalienceTransformerDecoderLayer(
            embed_dim=embed_dim,
            n_heads=num_heads,
            dropout=0.0,
            activation=nn.ReLU(inplace=True),
            n_levels=num_feature_levels,
            n_points=4,
            d_ffn=dim_feedforward,
        ),
        num_layers=transformer_dec_layers,
        num_classes=num_classes,
    ),
    num_classes=num_classes,
    num_feature_levels=num_feature_levels,
    two_stage_num_proposals=num_queries,
    level_filter_ratio=(0.4, 0.8, 1.0, 1.0),
    layer_filter_ratio=(1.0, 0.8, 0.6, 0.6, 0.4, 0.2),
)

matcher = HungarianMatcher(cost_class=2, cost_bbox=5, cost_giou=2)

weight_dict = {"loss_class": 1, "loss_bbox": 5, "loss_giou": 2}
weight_dict.update({"loss_class_dn": 1, "loss_bbox_dn": 5, "loss_giou_dn": 2})
weight_dict.update({
    k + f"_{i}": v
    for i in range(transformer_dec_layers - 1)
    for k, v in weight_dict.items()
})
weight_dict.update({"loss_class_enc": 1, "loss_bbox_enc": 5, "loss_giou_enc": 2})
weight_dict.update({"loss_salience": 2})

criterion = HybridSetCriterion(num_classes, matcher=matcher, weight_dict=weight_dict, alpha=0.25, gamma=2.0)
foreground_criterion = SalienceCriterion(noise_scale=0.0, alpha=0.25, gamma=2.0)
postprocessor = PostProcess(select_box_nums_for_evaluation=300)

# combine above components to instantiate the model
model = SalienceDETR(
    backbone=backbone,
    neck=neck,
    position_embedding=position_embedding,
    transformer=transformer,
    criterion=criterion,
    focus_criterion=foreground_criterion,
    postprocessor=postprocessor,
    num_classes=num_classes,
    num_queries=num_queries,
    aux_loss=True,
    min_size=800,
    max_size=1333,
)

```

**Comment by phuongdnm** [comment]
@xiuqhou Thank you so much for your guidance.
I will share the result with you when I complete my implementation and training.

**Comment by phuongdnm** [comment]
@xiuqhou 
I have completed my training successfully. I used the backbone model "mobilenetv4_conv_aa_large" as you suggested.
However, when I run the test script to evaluate my model, I got 0.0 results for all benchmark (image below).

This is the command I used for testing
```
CUDA_VISIBLE_DEVICES=6,7 accelerate launch test.py --coco-path data --model-config configs/salience_detr/salience_detr_mobilenet_800_1333.py --checkpoint model_zoo/best_ap50.pth
```

Do you know which step I did wrong during my testing?

![Image](https://github.com/user-attachments/assets/c27ad38e-7ff1-415b-bf82-e7c2a121363d)

**Comment by phuongdnm** [comment]
@xiuqhou also, when I tried the inference test with the trained model, it does not annotate any bounding boxes.
I believe I did something incorrectly.

![Image](https://github.com/user-attachments/assets/c1f463c5-5f64-4cf9-95a0-b3149205916e)

**Comment by phuongdnm** [comment]
@xiuqhou Hi, can you please check my 2 comments above?
Thank you in advance.

**Comment by xiuqhou** [comment]
Hi @phuongdnm , sorry for the delay. 
I didn't find mistakes in the configs. I guess a lightweight backbone may need more epochs to converge. Have you tried to train more epochs to see if AP has changed?

I will train myself with this config to see if AP changes correctly.

**Comment by phuongdnm** [comment]
@xiuqhou Thank you for your response.
I check the training log and it appears that the model has mAP=42.2 after the 12 epochs.

Then it saves the model in to log folder:
```
[2025-01-19 14:01:42 accelerate.tracking] DEBUG: Successfully logged to TensorBoard
[2025-01-19 14:01:51 det.util.utils] INFO: the best ap model is saved to checkpoints/salience_detr_mobilenet_800_1333/train/2025-01-18-10_51_55/best_ap.pth
[2025-01-19 14:01:59 det.util.utils] INFO: the best ap50 model is saved to checkpoints/salience_detr_mobilenet_800_1333/train/2025-01-18-10_51_55/best_ap50.pth
[2025-01-19 14:01:59 det.__main__] INFO: Training time: 1 day, 3:09:53
[2025-01-19 14:01:59 accelerate.tracking] DEBUG: TensorBoard writer closed
```

But when I execute the test script (in my previous [message](https://github.com/xiuqhou/Salience-DETR/issues/59#issuecomment-2599406007)), it turns out to mAP=0 😢
If you want me to share the pre-trained model, please let me know.

**Comment by xiuqhou** [comment]
@phuongdnm  Then I think there is some mistakes when loading checkpoint. Can you check test output to see if there is some thing like `all keys matched successfully`:

![Image](https://github.com/user-attachments/assets/5d397fff-374e-4aa7-9a2e-b7c561121d2a)

**Comment by phuongdnm** [comment]
@xiuqhou I see. 
I rerun the test command and I also see the line `All keys matched successfully`.

<img width="1510" alt="Image" src="https://github.com/user-attachments/assets/d3576ce9-a619-44f6-ae44-c2d9b38e2e6c" />

This time, the test give me some different output, but mAP is still = 0.

<img width="1041" alt="Image" src="https://github.com/user-attachments/assets/ac00113c-b9ec-4b9d-9cf4-58335d28727e" />

**Comment by xiuqhou** [comment]
This means that the checkpoint has been correctly loaded. Then I guess the checkpoint itself may have some mistakes. What if you load other checkpoints, for example, `pytorch_model.bin` under the last checkpoint directory.

**Comment by phuongdnm** [comment]
@xiuqhou I see.
I have just tried `pytorch_model.bin` and `best_ap50.pth` checkpoints.
Unfortunately, I still achieved the same result as `best_ap.pth` (mAP = 0.000).

**Comment by xiuqhou** [comment]
Still can not figure out what's wrong 😧 It looks that training works well and the checkpoint is good. But loading checkpoint into the model produces incorrect test AP. 

I just did a train and test myself on a small dataset with the above `mobilenetv4` config. The test script produced AP the same as train. 

train:

![Image](https://github.com/user-attachments/assets/7cda3db6-e462-4bf0-81c5-c3282f54f744)

test:

![Image](https://github.com/user-attachments/assets/6521a213-d400-4342-90e2-56a5e4cc119b)

Can you please check if there is any code changed between training and testing, or different from original code in this repo?

**Comment by phuongdnm** [comment]
@xiuqhou thank you so much for checking.
I decided to retrain the model with 2 epochs to test this behavior.
I use the same config that you shared above.
However, the `test.py` script still outputs a different result compared to training process.
During training, mAP is around 0.075 after 2 epoch, but when testing, mAP is still 0.000.

![Image](https://github.com/user-attachments/assets/cfb43fc1-bf3d-4619-81ce-cb81302db553)

Can you please retrain the model on Coco dataset with just 1 or 2 epochs?
I'm using the latest code in the main branch.

**Comment by xiuqhou** [comment]
OK. To make sure we have the same settings and scripts, could you please send your modified code to my email? I will use your code to retrain the model on COCO and see what will happen. 

This is my email: xiuqhou@stu.xjtu.edu.cn

**Comment by phuongdnm** [comment]
@xiuqhou Please give me a moment.
Let me create a fork from your repository then copy my file to the fork.
After that, you can clone my fork repository to save the download time.

I will retrain the model using my fork as well!

**Comment by phuongdnm** [comment]
@xiuqhou Please use my fork here: https://github.com/phuongdnm/Salience-DETR
The only change I need to make before training is the location of Coco dataset: https://github.com/phuongdnm/Salience-DETR/blob/main/configs/train_config_mobilenet.py#L20

I have started the training successfully using this command 
```
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch main.py --config-file configs/train_config_mobilenet.py
```

![Image](https://github.com/user-attachments/assets/52d739e8-0504-4606-9c7f-e2b288afb456)

So please let me know if you need any further information!

**Comment by xiuqhou** [comment]
@phuongdnm Thanks! I will let you know once I get results.

**Comment by phuongdnm** [comment]
@xiuqhou My training process was completed, and I reproduced the issue.
This is the log of the training:

![Image](https://github.com/user-attachments/assets/01e7254e-99a7-4437-aa86-83d33976c0f0)

This is the log of the `test.py` script:
```
CUDA_VISIBLE_DEVICES=0,1,2,3 accelerate launch test.py --coco-path my_coco_data --model-config configs/salience_detr/salience_detr_mobilenet_800_1333.py --checkpoint checkpoints/salience_detr_mobilenet_800_1333/train/2025-01-21-16_31_00/best_ap.pth
```

![Image](https://github.com/user-attachments/assets/db6b7a61-6bd2-4151-911d-d6ddb2f2f554)

**Comment by xiuqhou** [comment]
@phuongdnm I used three GPUs and it's still training. Would you mind sending me the checkpoints you have trained?

**Comment by phuongdnm** [comment]
@xiuqhou Please access this Google Drive folder: https://drive.google.com/drive/folders/1Gz0tCJkZLn87kus2X9oTkbprgw0_m6OV?usp=sharing

I uploaded the full checkpoint folder so you can check any information you want.

**Comment by xiuqhou** [comment]
Hi @phuongdnm 
Thanks for your checkpoints. I have also reproduced the issue!
I found that the model weight saved by `accelerate` package is incorrect in DDP training. This problem disappeared when I used a single GPU to train the model. I will figure out how to fix it!

**Comment by xiuqhou** [comment]
@phuongdnm 
I finally found out the issue was caused by `convert_sync_batchnorm`. Some `timm` models use their defined `BatchNormAct2d` that contains `norm`, `drop` and `act`. When converted with `torch.nn.SyncBatchNorm.convert_sync_batchnorm`, `drop` + `act` are ignored. (https://github.com/huggingface/pytorch-image-models/issues/1254)

To solve the issue, you can replace `torch.nn.SyncBatchNorm.convert_sync_batchnorm` with their newly defined `timm.layers.norm_act.convert_sync_batchnorm` in `main.py`.

```python
# original conversion in `main.py`
model = torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)

# updated conversion
model = timm.layers.norm_act.convert_sync_batchnorm(model)
```

**Comment by phuongdnm** [comment]
@xiuqhou after checking the issue link above, I can understand the root cause.

I think I will use the new conversion to train 1 epoch with my previous checkpoint (https://github.com/xiuqhou/Salience-DETR/issues/59#issuecomment-2604556718).
After that, I will run the test script again and let you know the result.

If it does not work, I will retrain the model without checkpoint.

**Comment by phuongdnm** [comment]
@xiuqhou It works!
I started a training using 1 epoch with the previous `checkpoint/best_ap.pth`.
After that, when I execute the script, I got the correct result!

**Comment by phuongdnm** [comment]
@xiuqhou Thank you so much for your support! 
I have tested and both the checkpoint and training from epoch 0 works with the new updates.

I would like to ask you a last question. From your experience, which parts can I add more to Salience DETR to achieve higher performance after training?

**Comment by xiuqhou** [comment]
@phuongdnm Glad to hear it works!
To achieve higher performance after or in training, there are some suggestions:
1. Use larger backbone and more complicated data augmentation in `transforms/presets.py`.
2. Increase `num_queries`, `level_filter_ratio` and `layer_filter_ratio` in transformer , decrease `groups` in RepVGGPluXNetwork.
3. Add more architectural designs or NMS.

**Comment by phuongdnm** [comment]
@xiuqhou Thank you so much for your time and supports!
I will try your suggestion.

Since these modifications are not related to this issue anymore, I would like to close the issue.
I will create a new issue related to those topics if I found something new.
